kratos/modeler/clean_up_problematic_triangles_modeler.cpp

```diff
@@ -0,0 +1,306 @@
+
+    Timer::Stop("CleanUpProblematicTrianglesModeler");
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void CleanUpProblematicTrianglesModeler::CleanUpProblematicGeometriesInMesh(
+    ModelPart& rThisModelPart,
+    const std::string& rEntityType,
+    const IndexType FirstNodeId,
+    const IndexType FirstElementId,
+    const IndexType FirstConditionId,
+    const double AreaTolerance
+    )
+{
+    if (rEntityType == "element") {
+        CleanUpProblematicGeometries<Element>(rThisModelPart, FirstNodeId, FirstElementId, AreaTolerance);
+    } else if (rEntityType == "condition") {
+        CleanUpProblematicGeometries<Condition>(rThisModelPart, FirstNodeId, FirstConditionId, AreaTolerance);
+    } else if (rEntityType == "geometry") { // TODO: Implement geometry cleanup
+        KRATOS_ERROR << "Geometry cleanup not yet supported. Sorry" << std::endl;
+    } else {
+        KRATOS_ERROR << "Unknown entity type: " << rEntityType << std::endl;
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <typename TEntityType>
+void CleanUpProblematicTrianglesModeler::CleanUpProblematicGeometries(
+    ModelPart& rThisModelPart,
+    const IndexType FirstNodeId,
+    const IndexType FirstEntityId,
+    const double AreaTolerance
+    )
+{
+    // Get the entities container
+    auto& r_entities = EntitiesUtilities::GetEntities<TEntityType>(rThisModelPart);
+
+    // Initialize variables
+    double average_length = 0.0;
+    double ref_area = 0.0;
+    {
+        // Compute the largest length of all geometries and the sum of lengths
+        const double sum_length = block_for_each<SumReduction<double>>(r_entities, [](auto& rEntity) {
+            const double length = rEntity.GetGeometry().Length();
+            if (length > 0.0) {
+                return length;
+            } else {
+                return 0.0;
+            }
+        });
+        average_length = sum_length / static_cast<double>(r_entities.size());
+        ref_area = average_length * average_length * AreaTolerance;
+    }
+    ref_area = ref_area * ref_area; // Using squared area tolerance to avoid expensive square roots, also will avoid numerical issues due to geometries that are 3 nodes in the same line, which squared area is negative and therefore area will be undefined (NaN)
+
+    // Lambda function to check if two nodes are at the same position using relative tolerance
+    auto ComputeDistance = [](const Node& rNode1, const Node& rNode2) -> double {
+        const array_1d<double, 3>& r_coords1 = rNode1.Coordinates();
+        const array_1d<double, 3>& r_coords2 = rNode2.Coordinates();
+        return norm_2(r_coords1 - r_coords2);
+    };
```

Comment:
Maybe you want to add these lambdas to the `private` section of the header to have a less verbose implementation.

Reply:
Okay